
# Script to generate DNAseI track for hg38 genome
[link](https://genome.ucsc.edu/cgi-bin/hgTables?db=hg38&hgta_group=regulation&hgta_track=wgEncodeRegDnase&hgta_table=wgEncodeRegDnaseUwTh1Peak&hgta_doSchema=describe+table+schema)


````{r}
rm(list = ls())
project_dir <- '/data/GenomicTrackRepository'
# Libraries
libraries <- list('dplyr')
invisible(suppressMessages(do.call(library, libraries)))
invisible(suppressMessages(sapply(list.files(file.path(project_dir, 'src', 'Rlib'), pattern = "utils.R", full.names = TRUE), source)))

project_paths <- build_project_paths(project_dir)
````

# Standard

```{r}

aavenger_sites <- vroom::vroom(file.path(project_dir, 'notebooks', 'temp', 'Rivella_20240516.tsv.gz')) %>%
  dplyr::mutate(sampleName = patient)

print(aavenger_sites %>% dplyr::filter(is.na(sampleName)) %>% nrow())

raw_count <- print(nrow(aavenger_sites))

aavenger_sites <- aavenger_sites %>%
  # dplyr::filter(is.na(start))
  dplyr::filter(!stringr::str_detect(seqnames, 'chrUn|_alt|_random')) %>% # removes 168 sites
  dplyr::mutate(
    trial = '',
    subject = sampleName,
    sample = '') %>%
  InvestigateIntegrations::format_aavenger_sites() %>%
  dplyr::filter(!is.na(start))

print(paste0('Number of sites removed: ', raw_count - nrow(aavenger_sites)))

chromosome_lengths <- InvestigateIntegrations::example_hg38_chromosome_lengths

random_match_df <- InvestigateIntegrations::aavenger_sites_random_match(
  aavenger_sites = aavenger_sites,
  chromosome_lengths = chromosome_lengths,
  random_seed_value = 10,
  match_row_number_modifier = 3
  )
  
print(head(random_match_df))

# combine random match and experimental dataframes
combined_df <- rbind(aavenger_sites, random_match_df)
```

## View integration site by chromosome counts

```{r}

df_percentage <- combined_df %>%
         dplyr::group_by(heatmap_group, type, seqname) %>%
         dplyr::summarize(count = dplyr::n()) %>%
         dplyr::mutate(percentage = 100 * (count/sum(count))) %>%
         dplyr::ungroup()

ggplot(combined_df %>%
         dplyr::group_by(heatmap_group, type, seqname) %>%
         dplyr::summarize(count = dplyr::n()) %>%
         dplyr::mutate(percentage = 100 * (count/sum(count))) %>%
         dplyr::ungroup(), 
       aes(x = type, y = percentage, fill = seqname))+
  facet_wrap(~heatmap_group) +
  geom_bar(stat = 'identity', position = position_stack())


dfx <- combined_df %>%
  dplyr::filter(heatmap_group == '_R003851 D14_', type == 'insertion', seqname == 'chr17') %>%
  dplyr::arrange(start)

df_results <- data.frame()
cluster_group <- 1
for (i in 1:nrow(dfx)) {
  
  if (i+1 > nrow(dfx)) {
    break
  }
  
  site_ <- dfx[i, ]
  
  next_site <- dfx[i+1, ]
  
  if (site_$start + 50000 > next_site$start - 50000) {
    
    df_results <- rbind(
      df_results,
      data.frame(
        'seed_site' = site_$start,
        'next_site' = next_site$start,
        'group' = cluster_group
      )
    )
    
  } else {
    cluster_group <- cluster_group + 1
  }
}

df_results2 <- df_results %>%
  dplyr::group_by(group) %>%
  dplyr::mutate(
    member_sites = length(next_site) + 1,
    start = min(seed_site),
    end = max(next_site),
    sites = paste(scales::comma(next_site), collapse = '-> '),
    site_distance = next_site - start,
    sites_distances_char = paste(scales::comma(site_distance), collapse = ' -> '),
    mean_distance = mean(site_distance)
    ) %>%
  dplyr::select(-c(seed_site, next_site, site_distance)) %>%
  dplyr::ungroup() %>%
  base::unique() %>%
  dplyr::mutate(
    group_range = end - start,
    group_range_char = scales::comma(group_range),
    ucsc_id = paste0('chr17:', start, '-', end)
  )

```


```{r}

# get the provided RData and rds feature files into a list
feature_files_to_process <- list.files(file.path(project_paths$data_processed, 'hg38'), pattern = "cpgIslandExt.rds|ncbiRefSeqCurated_transcription_start.rds|ncbiRefSeqCurated_transcription.rds|wgEncodeRegDnaseClustered.rds", full.names = TRUE)

print(feature_files_to_process)

# test each integration site for overlap in each feature at each given overlap 
combined_overlap_test_results_genomic_ranges <- InvestigateIntegrations::test_for_overlaps(
  matched_aavenger_sites_df = combined_df,
  list_of_feature_files = feature_files_to_process,
  overlap_ranges_to_test = c(0, 1000, 10000, 100000, 1000000)
)

print(head(combined_overlap_test_results_genomic_ranges))

```

# GC content

```{r}

if (!file.exists(file.path(project_dir, 'notebooks', 'temp', 'gc_calculation.rds'))) {
  
  genomic_sequence <- Biostrings::readDNAStringSet(file.path('/data/topo_scratch', 'hg38.fa.gz'))

  chromosomes_to_keep <- names(genomic_sequence)[!stringr::str_detect(names(genomic_sequence), '_random|_alt|chrUn|_fix')]
    
  genomic_sequence <- genomic_sequence[names(genomic_sequence) %in% chromosomes_to_keep]
  
  result <- GC_percentage_per_insertion(
    df_combined = combined_df,
    genomic_sequence = genomic_sequence,
    chromosome_lengths = InvestigateIntegrations::example_hg38_chromosome_lengths,
    window_sizes_to_test = c(100, 1000, 10000, 100000, 1000000),
    fasta_name = 'genome'
  )
  
  # saveRDS(result, file.path(project_dir, 'notebooks', 'temp', 'gc_calculation.rds'))
  
}

result <- readRDS(file.path(project_dir, 'notebooks', 'temp', 'gc_calculation.rds'))
  
result <- result %>%
  dplyr::select(-unique_id) %>%
  dplyr::filter(seqname != 'chrM') %>%
  dplyr::mutate(
    dplyr::across(
      c(where(is.numeric)),
      tidyr::replace_na,
      0
    )
    )

```


# Combine and use hotrocs

```{r}

all_genomic_tests <- merge(
  combined_overlap_test_results_genomic_ranges,
  result %>%
    dplyr::select(-c(width, strand, end, mid)),
  by = c('type', 'heatmap_group', 'seqname', 'start')
)


# manually change window to reflect that included in the filename
hot_roc_result_gc <- InvestigateIntegrations::hotroc_compare_insertion_to_match(
  matched_overlap_df = all_genomic_tests
  )

saveRDS(hot_roc_result_gc, file.path(project_dir, 'notebooks', 'temp', 'raw_hotrocs_result.rds'))
```

# reformatted

```{r, fig.asp=1, fig.width=8}

experimental_subject_order <- c("ALS17.1", "ALS17.2", "ALS17.3", "ALS20.1", "ALS20.2", "ALS20.3", "BB305.1", "BB305.2", "BB305.3",
                   "CHOP.1.D0", "CHOP.1.D14", "210410091.D0",  "210410091.D14","R003850.D0", "R003850.D14", "R003851.D0", "R003851.D14",
                   "WAS1", "WAS2", "WAS3", "CGD1", "CGD2", "CGD3", "CGD4")

hot_roc_result_mod <- InvestigateIntegrations::format_hot_roc_result(readRDS(file.path(project_dir, 'notebooks', 'temp', 'raw_hotrocs_result.rds')))


hot_roc_result_mod <- hot_roc_result_mod %>%
  dplyr::mutate(
    experimental_subject = stringr::str_replace_all(heatmap_group, 'X_', ''),
    experimental_subject = stringr::str_replace_all(experimental_subject, '_', ''),
    experimental_subject_factor = factor(
      experimental_subject, 
      levels = experimental_subject_order
      )
    ) %>%
  dplyr::mutate(
    window = sapply(feature_window, function(x) {stringr::str_split(x, '\\.')[[1]][3]}),
    window = as.integer(window),
    feature = sapply(feature_window, function(x) {stringr::str_split(x, '\\.')[[1]][1]}),
    feature_short = dplyr::case_when(
      feature == 'cpgIslandExt' ~ 'CpG',
      feature == 'ncbiRefSeqCurated_transcription' ~ 'RefSeqGene',
      feature == 'ncbiRefSeqCurated_transcription_start' ~ 'TSS',
      feature == 'wgEncodeRegDnaseClustered' ~ 'DNaseI',
      feature == 'genome_gc' ~ 'GC',
      TRUE ~ 'lookatme'
    ),
    feature_short_factor = factor(feature_short, c('DNaseI', 'CpG', 'GC', 'RefSeqGene', 'TSS'))
  ) %>%
  dplyr::arrange(feature_short_factor, dplyr::desc(window)) %>%
  dplyr::mutate(
    window_factor = dplyr::case_when(
      window == 0 ~ '0',
      window == 100 ~ '100',
      window == 1000 ~ '1K',
      window == 10000 ~ '10K',
      window == 100000 ~ '100K',
      window == 1000000 ~ '1M',
      TRUE ~ 'lookatme'
    ),
    window_factor = factor(window_factor, levels = rev(c('0', '100', '1K', '10K', '100K', '1M'))),
    # window_factor = factor(as.character(window), levels = base::unique(as.character(window))),
    feature_window_factor = factor(feature_window, levels = base::unique(feature_window)))


hot_roc_result_mod <- hot_roc_result_mod %>%
  dplyr::mutate(
    comparison_group = dplyr::case_when(
      experimental_subject %in% experimental_subject_order[1:9] ~ 'group1',
      experimental_subject %in% experimental_subject_order[10:17] ~ 'group2',
      experimental_subject %in% experimental_subject_order[18:24] ~ 'group3',
      TRUE ~ 'lookatme'
    ),
    comparison_group_factor = factor(comparison_group, levels = c('group1', 'group2', 'group3'))
  )

hot_roc_result_mod <- hot_roc_result_mod %>%
  dplyr::filter(
    !(window > 1000 & stringr::str_detect(feature_short_factor, 'TSS')),
    !(window < 1000 & stringr::str_detect(feature_short_factor, 'DNaseI')),
    !(window < 1000 & stringr::str_detect(feature_short_factor, 'CpG')),
  )
  
p1 <- ggplot(hot_roc_result_mod, aes(y = feature_window_factor, x = experimental_subject_factor, fill = round(ROC_value, 2))) +
  ggh4x::facet_nested(feature_short_factor + window_factor ~ comparison_group_factor, scales = 'free', space = 'free', switch = 'y') +
  geom_tile(color = 'black', width = 0.95, height = 0.95) + 
  geom_text(aes(label = sig_p_value), nudge_y = 0) +
  # geom_text(aes(label = round(ROC_value, 2)), nudge_y = -0.25, size = 2.5, color = 'black') +
  coord_cartesian(clip = 'off') +
  # colors
  # scale_fill_gradientn(colours = c("blue", "white", "red"), 
  #                      na.value = "transparent", 
  #                      breaks = c(0, 0.25, 0.5, 0.75, 1), 
  #                      labels = c(0, 0.25, 0.5, 0.75, 1), 
  #                      limits = c(0, 1)) +
  # scale_fill_gradientn(colours = c("blue", "white", "red"), 
  #                      na.value = "transparent", 
  #                      limits = c(0.35, 0.85),
  #                      breaks = c(0.35, 0.45, 0.5, 0.55, 0.65, 0.75, 0.85),
  #                      labels = c(0.35, 0.45, 0.5, 0.55, 0.65, 0.75, 0.85)) +
scale_fill_gradient2(
  low = "blue", high = "red", mid = "white",
  midpoint = 0.5, limit = c(0.4, 0.85)) +
  theme_classic() +
  theme(
    # aspect.ratio = 1,
    strip.text.y.left = element_text(angle = 0),
    strip.clip = 'off',
    axis.ticks.y = element_blank(),
    axis.text.y = element_blank(),
    axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
    axis.title = element_blank()
  )

cowplot::save_plot(
  filename = file.path('/data/GenomicTrackRepository/notebooks/temp', 'updated_example_heatmap.pdf'),
  plot = p1,
  base_asp = 0.5,
  base_width = 10,
  base_height = 15
)

```


# Chuck comparison

```{r}

dfx <- readRDS(file.path('/data/GenomicTrackRepository/notebooks/temp', 'roc.res.rds')) %>%
  InvestigateIntegrations::format_hot_roc_result()




```